### PR TITLE
Implement low-level interface moveit_cpp

### PR DESCRIPTION
**Pull request by henningkayser**
As already discussed in the last maintainer meeting we at PickNik have a need for a set of low-level interface classes that allow us to use, modify and exchange core components without the overhead 
 to adjust message types. Also, we are implementing performance-critical functions that are not throttled by the ROS ecosystem. The first MVP is a MoveGroup-like set of features that allows basic planning and execution for different planning groups. I will follow-up with more detailed explanation of the goals, requirements, roadmap and design decisions.

### CODE REVIEW

File: moveit_ros/planning_interface/moveit_cpp/src/moveit.cpp [hunk 1]
```diff
@@ -0,0 +1,335 @@
+/*********************************************************************
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit.cpp]: moveit_cpp.cpp

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/transform_listener.h>
+
+namespace moveit
+{
+namespace planning_interface
+{
+MOVEIT_CLASS_FORWARD(MoveitCpp);
+
+class MoveitCpp
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]:
```suggestion
class MoveItCpp
```

File: moveit_ros/planning_interface/moveit_cpp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+set(MOVEIT_LIB_NAME moveit_cpp)
+
+add_library(${MOVEIT_LIB_NAME} src/moveit_cpp.cpp src/planning_component.cpp)
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/CMakeLists.txt]: list on individual lines

File: moveit_ros/planning_interface/moveit_cpp/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
+target_link_libraries(${MOVEIT_LIB_NAME} moveit_common_planning_interface_objects moveit_planning_scene_interface ${catkin_LIBRARIES} ${Boost_LIBRARIES})
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/CMakeLists.txt]: individual lines

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Henning Kayser */
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: Add description of what this file does

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    }
+    std::string name;
+    std::string robot_description;
+    std::string joint_state_topic;
+    std::string attached_collision_object_topic;
+    std::string monitored_planning_scene_topic;
+    std::string publish_planning_scene_topic;
+    double wait_for_initial_state_timeout;
+  };
+  struct PlanningPipelineOptions
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: Add desc comment

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    void load(const ros::NodeHandle& nh)
+    {
+      std::string ns = "planning_pipelines/";
+      nh.getParam(ns + "pipeline_names", pipeline_names);
+      nh.getParam(ns + "namespace", parent_namespace);
+    }
+    std::vector<std::string> pipeline_names;
+    std::string parent_namespace;
+  };
+  struct PlannerOptions
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: Add desc comment

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      nh.getParam(ns + "planning_attempts", planning_attempts);
+      nh.getParam(ns + "max_velocity_scaling_factor", max_velocity_scaling_factor);
+      nh.getParam(ns + "max_acceleration_scaling_factor", max_acceleration_scaling_factor);
+    }
+    int planning_attempts;
+    double planning_time;
+    double max_velocity_scaling_factor;
+    double max_acceleration_scaling_factor;
+  };
+
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: Add blank line between above structs, just like here

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    PlanningSceneMonitorOptions planning_scene_monitor_options;
+    PlanningPipelineOptions planning_pipeline_options;
+    PlannerOptions default_planner_options;
+  };
+
+  /** \brief Constructor */
+  MoveItCpp(const ros::NodeHandle& nh, const std::shared_ptr<tf2_ros::Buffer>& tf_buffer = {});
+  MoveItCpp(const Options& opt, const ros::NodeHandle& nh, const std::shared_ptr<tf2_ros::Buffer>& tf_buffer = {});
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: opt -> options

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+  /** \brief Get the RobotModel object. */
+  robot_model::RobotModelConstPtr getRobotModel() const;
+
+  /** \brief Get the ROS node handle of this instance operates on */
+  const ros::NodeHandle& getNodeHandle() const;
+
+  /** \brief Get the current state queried from the current state monitor */
+  bool getCurrentState(robot_state::RobotStatePtr& current_state, double wait_seconds);
+
+  /** \brief Get the current state queried from the current state monitor */
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: document wait_seconds behavior

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+  /** \brief Destructor */
+  ~MoveItCpp();
+
+  /** \brief Get the RobotModel object. */
+  robot_model::RobotModelConstPtr getRobotModel() const;
+
+  /** \brief Get the ROS node handle of this instance operates on */
+  const ros::NodeHandle& getNodeHandle() const;
+
+  /** \brief Get the current state queried from the current state monitor */
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: document wait_seconds behavior

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+               bool blocking = true);
+
+protected:
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+
+private:
+  //  Core properties and instances
+  ros::NodeHandle node_handle_;
+  std::string robot_description_;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h]: i don't think we need this var, remove

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+{
+  ROS_INFO_NAMED(LOGNAME, "Deleting MoveItCpp");
+  clearContents();
+}
+
+MoveItCpp& MoveItCpp::operator=(MoveItCpp&& other)
+{
+  if (this != &other)
+  {
+    this->robot_description_ = other.robot_description_;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: i don't think we need this var, remove

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+const std::shared_ptr<tf2_ros::Buffer>& MoveItCpp::getTFBuffer() const
+{
+  return tf_buffer_;
+}
+
+void MoveItCpp::clearContents()
+{
+  robot_description_.clear();
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: i don't think we need this var, remove

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Henning Kayser */
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]: Add desc to this file... what does it do?

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+    }
+    MoveItErrorCode(const moveit_msgs::MoveItErrorCodes& code)
+    {
+      val = code.val;
+    }
+    explicit operator bool() const
+    {
+      return val == moveit_msgs::MoveItErrorCodes::SUCCESS;
+    }
+    bool operator==(const int c) const
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]: c -> code

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+    {
+      return val == c;
+    }
+    bool operator!=(const int c) const
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]: c -> code

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+  /// The representation of a plan solution
+  struct PlanSolution
+  {
+    /// The full starting state used for planning
+    moveit_msgs::RobotState start_state;
+
+    /// The trajectory of the robot (may not contain joints that are the same as for the start_state_)
+    robot_trajectory::RobotTrajectoryPtr trajectory;
+
+    /// Error code
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]:
```suggestion
    /// Reason why the plan failed
```

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+    MoveItErrorCode error_code;
+
+    explicit operator bool() const
+    {
+      return bool(error_code);
+    }
+  };
+
+  /// Planner parameters provided with the MotionPlanRequest
+  struct PlanRequestParameters
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]: Why is the PlannerOptions struct so similar to this struct? Do we really need both? They have the same contents. Let's remove one.

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+  PlanningComponent& operator=(const PlanningComponent&) = delete;
+
+  PlanningComponent(PlanningComponent&& other);
+  PlanningComponent& operator=(PlanningComponent&& other);
+
+  /** \brief Destructor */
+  ~PlanningComponent();
+
+  /** \brief Get the name of the planning group */
+  const std::string& getName() const;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]:
```suggestion
  const std::string& getPlanningGroupName() const;
```

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+
+  /** \brief Get the names of the named robot states available as targets */
+  const std::vector<std::string> getNamedTargets();
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]:
```suggestion
  const std::vector<std::string> getNamedTargetStates();
```

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+
+  /** \brief Get the joint values for targets specified by name */
+  std::map<std::string, double> getNamedTargetValues(const std::string& name);
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]:
```suggestion
  std::map<std::string, double> getNamedTargetStateValues(const std::string& name);
```

File: moveit_ros/planning_interface/moveit_cpp/src/planning_component.cpp [hunk 23]
```diff
@@ -0,0 +1,354 @@
...
+#include <ros/console.h>
+#include <ros/ros.h>
+
+namespace moveit
+{
+namespace planning_interface
+{
+constexpr char LOGNAME[] = "planning_component";
+
+PlanningComponent::PlanningComponent(const std::string& group_name, const MoveItCppPtr& moveit_context)
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/planning_component.cpp]:
```suggestion
PlanningComponent::PlanningComponent(const std::string& group_name, const MoveItCppPtr& moveit_cpp)
```
consistent var name?

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+  bool execute(bool blocking = true);
+
+  /** \brief Return the last plan solution*/
+  const PlanSolutionPtr getLastPlanSolution();
+
+private:
+  // Core properties and instances
+  ros::NodeHandle nh_;
+  MoveItCppPtr moveit_cpp_;
+  const std::string group_name_;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]: do we need to store this string, or can we just use joint_model_group_->getName()?

File: moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+  // The robot_model_ member variable of MoveItCpp class will manually free the joint_model_group_ resources
+  const moveit::core::JointModelGroup* joint_model_group_;
+
+  // Planning
+  std::set<std::string> planning_pipeline_names_;
+  robot_state::RobotStatePtr considered_start_state_;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h]: Add comment about what 'considered' means

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 26]
```diff
@@ -0,0 +1,333 @@
...
+constexpr char LOGNAME[] = "moveit_cpp";
+constexpr char PLANNING_SCENE_MONITOR_NAME[] = "moveit_cpp_planning_scene";
+constexpr char PLANNING_PLUGIN_PARAM[] = "planning_plugin";
+
+MoveItCpp::MoveItCpp(const ros::NodeHandle& nh, const std::shared_ptr<tf2_ros::Buffer>& tf_buffer)
+  : MoveItCpp(Options(nh), nh, tf_buffer)
+{
+}
+
+MoveItCpp::MoveItCpp(const Options& opt, const ros::NodeHandle& nh, const std::shared_ptr<tf2_ros::Buffer>& tf_buffer)
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]:
```suggestion
MoveItCpp::MoveItCpp(const Options& options, const ros::NodeHandle& nh, const std::shared_ptr<tf2_ros::Buffer>& tf_buffer)
```

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 27]
```diff
@@ -0,0 +1,333 @@
...
+{
+  if (!tf_buffer_)
+    tf_buffer_.reset(new tf2_ros::Buffer());
+  tf_listener_.reset(new tf2_ros::TransformListener(*tf_buffer_));
+
+  // Configure planning scene monitor
+  if (!loadPlanningSceneMonitor(opt.planning_scene_monitor_options))
+  {
+    std::string error = "Unable to configure planning scene monitor";
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]:
```suggestion
    const std::string error = "Unable to configure planning scene monitor";
```

Comment (davetcoleman) [reply moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: Apply to strings below, also

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 28]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+MoveItCpp::MoveItCpp(MoveItCpp&& other)
+{
+  other.clearContents();
+}
+
+MoveItCpp::~MoveItCpp()
+{
+  ROS_INFO_NAMED(LOGNAME, "Deleting MoveItCpp");
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: Is this logger necessary?

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 29]
```diff
@@ -0,0 +1,333 @@
...
+    planning_scene_monitor_->startSceneMonitor(opt.monitored_planning_scene_topic);
+  }
+  else
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Planning scene not configured");
+    return false;
+  }
+  // TODO(henningkayser): fix and remove lines below
+  ros::spinOnce();
+  ros::Duration(0.5).sleep();  // when at 0.1, i believe sometimes vjoint not properly loaded
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: is this comment old from my boilerplate code? it looks familar to me. i think we should remove the sleep.

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 30]
```diff
@@ -0,0 +1,333 @@
...
+      }
+    }
+  }
+
+  return true;
+}
+
+robot_model::RobotModelConstPtr MoveItCpp::getRobotModel() const
+{
+  ROS_DEBUG_NAMED(LOGNAME, "MoveItCpp::getRobotModel()");
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: remove these loggers?

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 31]
```diff
@@ -0,0 +1,333 @@
...
+  return robot_model_;
+}
+
+const ros::NodeHandle& MoveItCpp::getNodeHandle() const
+{
+  ROS_DEBUG_NAMED(LOGNAME, "MoveItCpp::getNodeHandle()");
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: remove these loggers.. this is a simple getter

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp [hunk 32]
```diff
@@ -0,0 +1,333 @@
...
+    }
+    result_names.insert(pipeline_name);
+  }
+  return result_names;
+}
+
+const planning_scene_monitor::PlanningSceneMonitorPtr& MoveItCpp::getPlanningSceneMonitor() const
+{
+  return planning_scene_monitor_;
+}
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: add line break

File: moveit_ros/planning_interface/test/moveit_cpp_test.cpp [hunk 33]
```diff
@@ -0,0 +1,178 @@
...
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Jafar Abdi
+   Desc: Test the MoveItCpp and PlanningComponent interfaces
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/moveit_cpp_test.cpp]: Love that there's a Desc here!

File: moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp (hunk 29 above)
Comment (davetcoleman) [reply moveit_ros/planning_interface/moveit_cpp/src/moveit_cpp.cpp]: ping: remove the sleep

**Comment by davetcoleman** [comment]
``
/root/ros_ws/src/moveit/moveit_ros/planning_interface/moveit_cpp/include/moveit/moveit_cpp/planning_component.h:191:21: warning: ‘moveit::planning_interface::PlanningComponent::group_name_’ will be initialized after [-Wreorder]
   const std::string group_name_;
``

**Comment by JafarAbdi** [comment]
The last commit fix the kinetic error, please [see](https://travis-ci.org/ros-planning/moveit/jobs/614677322?utm_medium=notification&utm_source=github_status)

**Comment by henningkayser** [comment]
Cleaned up commit history, lgtm

**Comment by davetcoleman** [comment]
Hurray! v0 of moveit_cpp has been merged. @henningkayser can you announce this on MoveIt Discourse?

**Comment by rhaschke** [comment]
@JafarAbdi, @henningkayser the [new unit test](https://github.com/ros-planning/moveit/blob/0924e99de84783a3a4cae4a57e51017d5cc48a59/moveit_ros/planning_interface/test/moveit_cpp_test.test) is flaky and breaks Travis randomly. Please fix asap.

**Comment by JafarAbdi** [comment]
@rhaschke Thank you for reporting! I'm working on it right now! 

I just tried to run the test forever and indeed sometimes I get seg fault
```bash
ros.moveit_ros_planning_interface.moveit_cpp: MoveItCpp running
Segmentation fault (core dumped)
```

**Comment by henningkayser** [comment]
@rhaschke @v4hn We've debugged this and @JafarAbdi will file a PR for this. The issue is caused by the fake controller not publishing valid joint states for joints not configured in the initial state. In this particular case `panda_moveit_config` had an initial state `ready` for the full group which didn't include finger joints, leading to undefined values being published. The PR will fix this by publishing 0 values by default.

**Pull request merged**